### ISSUE TITLE
The performance of fp16  is quite bad.

### ISSUE THREAD

**Issue by AlvinChen13** [tensorflow/tensorflow#5592]
https://github.com/tensorflow/tensorflow/issues/1300

I have got a nvidia p100 GPU which is support fp16, and I run the TF case 'cifar10_train.py'. Without option '--use_fp16', the performance is also 1600 examples/sec, and with the option '--use_fp16', the performance down to 500 examples/sec. Any ideas about this issue?

userid@ubuntu-WK-4xP100:~/weike/tensorflow-r0.11/tensorflow/models/image/cifar10$ vi cifar10_train.py 
userid@ubuntu-WK-4xP100:~/weike/tensorflow-r0.11/tensorflow/models/image/cifar10$ python cifar10_train.py 
I tensorflow/stream_executor/dso_loader.cc:111] successfully opened CUDA library libcublas.so.8.0 locally
I tensorflow/stream_executor/dso_loader.cc:111] successfully opened CUDA library libcudnn.so.5 locally
I tensorflow/stream_executor/dso_loader.cc:111] successfully opened CUDA library libcufft.so.8.0 locally
I tensorflow/stream_executor/dso_loader.cc:111] successfully opened CUDA library libcuda.so.1 locally
I tensorflow/stream_executor/dso_loader.cc:111] successfully opened CUDA library libcurand.so.8.0 locally
>> Downloading cifar-10-binary.tar.gz 100.0%
...
2016-11-14 00:07:57.143739: step 0, loss = 4.67 (12.1 examples/sec; 10.549 sec/batch)
2016-11-14 00:07:58.395209: step 10, loss = 4.57 (1693.6 examples/sec; 0.076 sec/batch)
2016-11-14 00:07:59.177525: step 20, loss = 4.97 (1668.9 examples/sec; 0.077 sec/batch)
2016-11-14 00:07:59.957789: step 30, loss = 4.43 (1588.3 examples/sec; 0.081 sec/batch)
2016-11-14 00:08:00.738431: step 40, loss = 4.52 (1690.5 examples/sec; 0.076 sec/batch)
2016-11-14 00:08:01.501940: step 50, loss = 4.33 (1680.4 examples/sec; 0.076 sec/batch)
2016-11-14 00:08:02.241604: step 60, loss = 4.20 (1733.4 examples/sec; 0.074 sec/batch)
2016-11-14 00:08:03.001845: step 70, loss = 4.27 (1706.0 examples/sec; 0.075 sec/batch)
2016-11-14 00:08:03.765522: step 80, loss = 4.18 (1601.3 examples/sec; 0.080 sec/batch)
2016-11-14 00:08:04.516780: step 90, loss = 4.25 (1646.3 examples/sec; 0.078 sec/batch)

userid@ubuntu-WK-4xP100:~/weike/tensorflow-r0.11/tensorflow/models/image/cifar10$ python cifar10_train.py --use_fp16
I tensorflow/stream_executor/dso_loader.cc:111] successfully opened CUDA library libcublas.so.8.0 locally
I tensorflow/stream_executor/dso_loader.cc:111] successfully opened CUDA library libcudnn.so.5 locally
I tensorflow/stream_executor/dso_loader.cc:111] successfully opened CUDA library libcufft.so.8.0 locally
I tensorflow/stream_executor/dso_loader.cc:111] successfully opened CUDA library libcuda.so.1 locally
I tensorflow/stream_executor/dso_loader.cc:111] successfully opened CUDA library libcurand.so.8.0 locally
Filling queue with 20000 CIFAR images before starting to train. This will take a few minutes.
..
2016-11-14 00:09:09.382854: step 0, loss = 4.67 (12.3 examples/sec; 10.411 sec/batch)
2016-11-14 00:09:11.923842: step 10, loss = 4.58 (659.6 examples/sec; 0.194 sec/batch)
2016-11-14 00:09:13.918448: step 20, loss = 6.62 (460.6 examples/sec; 0.278 sec/batch)
2016-11-14 00:09:16.211809: step 30, loss = 4.37 (583.7 examples/sec; 0.219 sec/batch)
2016-11-14 00:09:18.327690: step 40, loss = 4.30 (618.3 examples/sec; 0.207 sec/batch)
2016-11-14 00:09:20.395409: step 50, loss = 4.37 (643.8 examples/sec; 0.199 sec/batch)
2016-11-14 00:09:22.466230: step 60, loss = 4.32 (574.0 examples/sec; 0.223 sec/batch)
2016-11-14 00:09:24.533225: step 70, loss = 4.17 (646.2 examples/sec; 0.198 sec/batch)
2016-11-14 00:09:26.609277: step 80, loss = 2.59 (601.5 examples/sec; 0.213 sec/batch)
2016-11-14 00:09:28.703648: step 90, loss = 4.15 (625.5 examples/sec; 0.205 sec/batch)

**Comment by girving** [comment]
@benoitsteiner Would additional information be helpful?

**Comment by benoitsteiner** [comment]
@AlvinChen13 Did you compile TensorFlow yourself ? If so, which architecture did you target ?

**Comment by AlvinChen13** [comment]
@benoitsteiner Yes, I compile r0.11 with cuda 8, and my system is x86_64 with Tesla P100. And I just follow the guideline step by step to build TF. 

# To build with GPU support:
$ bazel build -c opt --config=cuda //tensorflow/tools/pip_package:build_pip_package

$ bazel-bin/tensorflow/tools/pip_package/build_pip_package /tmp/tensorflow_pkg

# The name of the .whl file will depend on your platform.
$ sudo pip install /tmp/tensorflow_pkg/tensorflow-0.11.0rc0-py2-none-any.whl

Any special options need be given while compiling TF to support fp16?

**Comment by skye** [comment]
Automatically closing due to lack of recent activity. Since this issue is old at this point, please reopen the issue if it still occurs when tried with the latest version of Tensorflow. Thank you.

**Comment by perone** [comment]
Any news on that ? I have the same issue with the P100, with TensorFlow 1.3.0 (tensorflow-gpu) installed from pip.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by rajivkapoor** [comment]
It has been a while since I revisited cifar10 training in fp16.  
I recently tried cifar10 training with --use_fp16 with a P100 using TF 1.4 and the perf is really bad compared to fp32.  
Some observations:

1. On my P100 fp32 trains at ~16000 examples/sec, while using --use_fp16 it drops down to ~780 examples/sec
2. I notice that for fp16 the GPU is highly underutilized ... just about 4% (as seen with nvidia-smi) and the CPU cores are pretty busy at ~80%. - so it seems like most ops are being done on the CPU.  
3. For fp32 training the CPU is about 30% busy and the GPU varies between 80 and 100%

Anyone else have better luck with this?

**Comment by quaeler** [comment]
#15643 seems like a ~duplicate of this

**Comment by KillPinguin** [comment]
I'm the author of #15585 and would like to comment on this / state a short version of my findings. Please see my issue for more.

@quaeler 
You're right, this could be a similar issue.

@AlvinChen13 could you try to profile your cifar example using nvprof? Like I stated in my issue, it seems that Tensorflow is using slower CUDA functions (*maxwell_fp16_segmemm_fp16_128x128_nn* for FP16 which is worse than the cuBLAS method *maxwell_hgemm_256x128_nn*)

@rajivkapoor I did not (yet) measure the CPU, since the actual GPU performance for elementary operations is worse... Could you also try running nvprof and check if the "raw" GPU performance drops when using FP16?

**Comment by tensorflowbutler** [comment]
Nagging Awaiting TensorFlower: It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Awaiting TensorFlower: It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Awaiting TensorFlower: It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging TensorFlower: It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @jart: It has been 28 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by bhack** [comment]
See also https://github.com/tensorflow/benchmarks/issues/77

**Comment by tensorflowbutler** [comment]
Nagging Assignee @jart: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by jart** [comment]
Thank you @bhack for sharing that link. That looks like a very authoritative answer. Please also note recent TensorFlow binary releases enable AVX by default. Things should be significantly better than they were back in 2016 when this issue was originally filed.

**Comment by rajivkapoor** [comment]
Better late than never :).  Recent TF builds (1.8) with CUDA 9.1 has indeed fixed the issue.  Depending on the batch size, I see pretty good fp16 perf. on NVidia P100 and V100 for both inference and training.
\rajiv

**Comment by visionscaper** [comment]
@rajivkapoor Does the batch size typically need to be higher to get good performance when training with fp16? How much bigger?

**Comment by rajivkapoor** [comment]
@Freddy Snijder - Strictly speaking about the TF CNN benchmarks, generally,
for a single V100 a bs >=8, fp16 gives pretty decent gains over fp32 - at
least for vgg11, inception3 and resnet*.
I have seen gains starting at bs >=4 for 2 or more GPUs per server.

**Comment by bckpkol** [comment]
Excuse me for the interruption, but float16 is also slow on Core2 CPU.

**Comment by Raghava14** [comment]
Hi,
If I trained a model on fp16 on a turing gpu and I want to do the inference on cpu is it possible. Would the inference time be faster and does Tensorflow support such senario because as I see PyTorch doesn't support fp16 inference on cpu. Is support on fp16 a limitation of intel cpus or the framework( like Tensorflow , PyTorch etc) .

**Comment by IvyGongoogle** [comment]
meet the same problem on v100: fp16 is very slow than fp32

**Comment by mttbx** [comment]
Met the same problem, fp16 is much slower than fp32

**Comment by bhateharsh** [comment]
Could you please provide us with a timeline of when the issue is expected to be resolved? I've faced the same problem on Jetson Nano as well as Laptop running CUDA 10, GTX 1050Ti.

**Comment by monklof** [comment]
meet the same problem on v100 with Cuda 9.0: fp16 is very slow than fp32.
